homeassistant/components/mqtt/__init__.py

```diff
@@ -852,7 +852,8 @@ def _availability_setup_from_config(self, config):
         """(Re)Setup."""
         self._availability_topic = config.get(CONF_AVAILABILITY_TOPIC)
         self._availability_qos = config.get(CONF_QOS)
-        self._available = self._availability_topic is None  # type: bool
+        if self._availability_topic is None:
+            self._available = True
```

Comment:
This doesn't looks like a fix. It's a rewrite to the long style for the condition.

```python
>>> a = None
>>> b = a is None
>>> b
True
>>> if a is None:
...     b = True
... 
>>> b
True
```

Reply:
and what about when `a` is not `None` in your example?
That's the fix :)

i.e. do not overwrite `self._available` if self._availability_topic is not None

Reply:
I agree this PR should have a better description. It's a subtle change that's pretty hard to follow without knowing the intricacies of discovery updates.

Let's consider this example: My MQTT discovered sensor is added using MQTT discovery and is using a retained availability topic `livingroom/status`. 

First, when the entity is added, the `available` property is set to `False` as we haven't heard the response from the availability subscription yet - so we don't know the available state.

Then, we get a discovery update with the same payload as before _or_ with something other than availability changed. We will **not** re-subscribe to the topic as it hasn't changed - but this check here is resetting the `available` property to `False` nonetheless. So the entity will remain unavailable even though there is a retained `online` payload on the availability topic the whole time.

So the change is: Only reset the available property to `True` if the availability topic feature has been removed.
The more correct version would be: reset iff the available options have **changed** - but that would be a complete mess. We would not only have to check if the topic has changed, but also if `qos`, `payload_available`, `payload_not_available` - and that all for a scenario which will probably never happen.

The scenario that there's a discovery update and the available options haven't changed will just be 100X more likely. So I think this fix is ok.

Two things to make this code cleaner:
 * The logic here and in the constructor is essentially the same. Why don't we pass `config` to the constructor and call `_availability_setup_from_config` from there? DRY
 * To keep this whole thing more sane, I would suggest moving the `self._availability_topic is None` logic to another place:

```python
@property
def available(self) -> bool:
    return self._availability_topic is None or self._available

# In constructor
self._available = False

# In _availability_setup_from_config, do not update self._available at all
```

Reply:
Thanks a lot @OttoWinter , I'll update as you suggest!